react/lib/gridstack-context.tsx

```diff
@@ -1,61 +1,91 @@
-import "gridstack/dist/gridstack-extra.css";
-import "gridstack/dist/gridstack.css";
-import type { ItemRefType } from "./gridstack-item";
+import { GridStack, GridStackWidget,GridStackOptions } from 'gridstack';
+import React, {
+  createContext,
+  PropsWithChildren,
+  useCallback,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+import isEqual from 'react-fast-compare';
 
 type GridStackContextType = {
-  grid: GridStack | null | undefined;
-  setGrid: React.Dispatch<React.SetStateAction<GridStack | null>>;
-  addItemRefToList: (id: string, ref: ItemRefType) => void;
-  removeItemRefFromList: (id: string) => void;
-  itemRefList: ItemRefListType;
-  getItemRefFromListById: (id: string) => ItemRefType | null;
+  getWidgetContent: (widgetId: string) => HTMLElement | null;
 };
 
-type ItemRefListType = {
-  id: string;
-  ref: ItemRefType;
-}[];
+interface GridstackProviderProps extends PropsWithChildren {
+  options: GridStackOptions;
+}
+
+export const GridstackContext = createContext<GridStackContextType | null>(null);
+
+export const GridstackProvider = ({ children, options }: GridstackProviderProps) => {
+  const widgetContentRef = useRef<Record<string, HTMLElement>>({});
+  const containerRef = useRef<HTMLDivElement>(null);
+  const optionsRef = useRef<GridStackOptions>(options);
 
-export const GridstackContext = React.createContext<GridStackContextType | null>(null);
+  const [parentGrid, setParentGrid] = useState<GridStack | null>(null);
 
-export const GridstackProvider = ({ children }: { children: React.ReactNode }) => {
-  const [grid, setGrid] = React.useState<GridStack | null>(null);
-  const [itemRefList, setItemRefList] = React.useState<ItemRefListType>([]);
+  const renderCBFn = useCallback((element: HTMLElement, widget: GridStackWidget) => {
+    if (widget.subGridOpts) {
+      element.classList.add('inset-0');
+      return;
+    }
 
-  const addItemRefToList = React.useCallback((id: string, ref: ItemRefType) => {
-    setItemRefList((prev) => [...prev, { id, ref }]);
+    if (widget.id) {
+      widgetContentRef.current[widget.id] = element;
+    }
   }, []);
 
-  const removeItemRefFromList = React.useCallback((id: string) => {
-    setItemRefList((prev) => prev.filter((item) => item.id !== id));
+  const getWidgetContent = useCallback((widgetId: string) => {
+    return widgetContentRef.current[widgetId] || null;
   }, []);
 
-  const getItemRefFromListById = React.useCallback((id: string) => {
-    const item = itemRefList.find((item) => item.id === id);
-    return item?.ref ?? null;
-  }, [itemRefList]);
+  const initGrid = useCallback(() => {
+    if (containerRef.current) {
+      GridStack.renderCB = renderCBFn;
+      return GridStack.init(optionsRef.current, containerRef.current);
+    }
+    return null;
+  }, [renderCBFn]);
+
+  useLayoutEffect(() => {
+    if (!isEqual(options, optionsRef.current) && parentGrid) {
+      try {
+        parentGrid.removeAll(false);
```

Comment:
this are some of the weird part about React and GS not being in sync and why in my angular wrapper I do NOT create the grid or items in html template but let GS call me back t create the appropriate component (based on NgWidgetItem.selector)

Reply:
This is for when the top options change we are destroying the current grid and reinitialising the same 
one of the thing why we are not letting GS dom manipulation is that React has their own dom manipulation logic 
so if we are interfering with the same things are not working as expected because as the grid element is already removed by react the the internal GS does not expect the same to happen